pkg/collector/listeners/docker.go

```diff
@@ -381,22 +385,53 @@ func (s *DockerService) GetPorts() ([]int, error) {
 		return nil, fmt.Errorf("failed to inspect container %s", string(s.ID)[:12])
 	}
 
-	for p := range cInspect.NetworkSettings.Ports {
-		portStr := string(p)
-		if strings.Contains(portStr, "/") {
-			portStr = strings.Split(portStr, "/")[0]
+	switch {
+	case cInspect.NetworkSettings != nil && len(cInspect.NetworkSettings.Ports) > 0:
+		for p := range cInspect.NetworkSettings.Ports {
+			out, err := parseDockerPort(p)
+			if err != nil {
+				log.Warn(err.Error())
+			} else {
+				ports = append(ports, out...)
+			}
 		}
-		port, err := strconv.Atoi(portStr)
-		if err != nil {
-			log.Warnf("failed to extract port %s", string(p))
+	case cInspect.Config != nil && len(cInspect.Config.ExposedPorts) > 0:
+		log.Infof("using ExposedPorts for container %s as no port bindings are listed", string(s.ID)[:12])
+		for p := range cInspect.Config.ExposedPorts {
+			out, err := parseDockerPort(p)
+			if err != nil {
+				log.Warn(err.Error())
+			} else {
+				ports = append(ports, out...)
+			}
 		}
-		ports = append(ports, port)
 	}
+
 	sort.Ints(ports)
 	s.Ports = ports
 	return ports, nil
 }
 
+func parseDockerPort(port nat.Port) ([]int, error) {
+	var output []int
+
+	first, last, err := port.Range()
+	if err == nil && last > first {
+		for p := first; p <= last; p++ {
+			output = append(output, p)
+		}
+		return output, nil
+	}
+
+	p := port.Int()
+	if p > 0 {
+		output = append(output, p)
+		return output, nil
+	}
+
+	return output, fmt.Errorf("failed to extract port %s", string(port))
```

Comment:
What's the value of port when we failed to extract it ?

Reply:
This should never happen as we get it from a docker object that is already checked on container start, but added it for support. Its internal type is `string`, hence the `%s`, but I moved to `%v` to be future-proof